### DISCUSSION TITLE
SeaweedFS Docker Swarm Stack

### DISCUSSION THREAD

**Discussion by prologic** [seaweedfs/seaweedfs#2086] [Show and tell]
HI All 👋

Just wanted to share this [Docker Swarm Stack](https://gist.github.com/prologic/d602c2c8f77ee8b9ccebab3db7183769) I put together and in the process of testing. So far things seem to work as expected.

I would love to hear any feedback anyone has on this setup and any improvements that _could_ be made.

The general idea here is to eventually have a SeaweedFS cluster either in a Swarm cluster of its own or run alongside (_i.e: hyperconverged_) your workload cluster(s) with node label affinity (e.g: `storage` labels).

Kind regards

James

**Comment by chrislusf** [comment]
Looks good. I am just a bit confused myself about docker-compose and docker swarm.

**Reply by prologic** [reply]
Thanks for your feedback and kind words. And thank you for building this very nicely designed, implemented and well documented file system!

So basically in a nutshell, `docker-compose` and the YAML spec it uses was born out of a desire to build a decent developer experience around docker containers. This was originally built by a 2-person startup company called Orchard which built a tool called `fig` where developers would typically define a YAML file and a set of Dockerfile(s) for images that make up their application stack and `fig up`. Much of this history can still be found [here](https://orchardup.github.io/fig.sh/)

Fast forward to where Docker Inc. (_at the time_) acquired Orchard and the `fig` codebase and talent, incorporated it into what later became known as `docker-compose` and `docker-compose.yml`.

Fast track a bit more where "Swarm Mode" became a first-class thing baked into the Docker Engine and you have `docker stack deploy -c weed.yml weed` (_for example_).

The key differences is Stacks (_as I and others call them_) are a superset of the original docker-compose/fig specs designed for deploy one or more services in a production cluster. `docker stack` these days actually understand how to deploy these stacks in both swarm and k8s clusters. The "orchestrator" was abstracted away at some point in the `docker stack` CLI and tooling.

**Reply by chrislusf** [reply]
Nice to learn these! This history should go to wikipedia. :)

This helps to guide the direction. Nowadays, k8s is much hotter than docker.

**Reply by prologic** [reply]
I'm glad you learned something 😀 I've been using Docker personally since v 0.7 or so way back when 🤣

> This helps to guide the direction. Nowadays, k8s is much hotter than docker.

I can't go into too much details due to security reasons, but one of the things we need in our organization is persistent storage. There are numerous "public cloud" solutions, many of which are geared towards k8s (_as you're aware_). None (_or very few_) I've found to date are as nicely designed, implemented and well documented as SeaweedFS. What's more, the Docker Volume Plugin you've also developed "just works"™ -- There are also numerous vendor-locked-in solutions as well like Portworx, Blockbridge and a few handful of others (_that may or may not work_) but are also quite expensive.

This is the only solution to our needs that give me the greatest hope/confidence.

This means of course we rely heavily on Docker Swarm (_and not Kubernetes_) so the maintenance of the docker volume plugin is important to us.

**Reply by chrislusf** [reply]
Good to hear the Docker Volume Plugin works. :) It was adapted from another file system and not well maintained. Definitely need some help with it.

**Reply by prologic** [reply]
I can probably lend a hand there 🤞

**Reply by jwillmer** [reply]
I'm currently looking to deploy our solutions in docker swarm since it offers the most simple configuration (learning curve) for a high availability setup. I am currently evaluating different distributed storage providers like: **storidge.com, portworx.com, storageos.com, gluster.org, min.io, beegfs.io** and of course **seaweedfs**.

@prologic can you share more details about your implementation? I'm at the start of the docker swarm journey although I have deep knowledge about docker(-compose). My first goal is to setup a cluster with 3 or 4 nodes, a distributed storage and traefik as revers proxy. You have any example config for scaling and binding the mounts to apps? Highly appreciate every bit I do not need figure out myself by try and error 😉

**Reply by prologic** [reply]
@jwillmer Sadly due to networking complexities I wasn't _actually_ able to get this Swarm deployment of SeaweedFS working correctly. I'm **quite** confident it can be done, but it needs work. It would help if the environment you create Docker Swarm clusters in had proper working DNS and each Node in the Swarm knew how to access it's external address (_as seen by the rest of the local network_).

That being said, I was able to get the Docker Volume plugin (_see my [fork](https://github.com/prologic/docker-plugin_seaweedfs)_) -- Unfortunately I ran into some problems that I do not believe have anything to do with the volume plugin (_as it directly fork/exec's `weed mount` and I was able to repro the same issues outside of Docker_).

FWIW Here are some of my own evaluations:

- StorageOS _no longer_ support Docker Swarm. Only Kubernetes :/
- Portworx is **super** expensive, but _supports_ both Swarm and Kubernetes.
- Storidge _honestly_ seems a bit of a joke. THeir setup fails in numerous places, their support is poor to non-existence and it seems like a storefront to something else (_great marketing material though_).

**Reply by jwillmer** [reply]
Had the same feeling about Storidge. I tried contacting them to get some direct response but seams like they don't care. I think [Minio also dropped support for Swarm](https://github.com/minio/minio/issues/12560) like StorageOS did. Seams like I need to switch to something like nomadproject.io

**Reply by prologic** [reply]
FWIW Nomad is _actually_ pretty great. I've used it quite successfully in the past. I am however heavily invested in Docker Swarm, so I will continue to invest time/effort where I can in solutions that work with Swarm.

**Reply by jwillmer** [reply]
Unexpectedly I got some reply from Storidge:

![image](https://user-images.githubusercontent.com/1503577/123167022-51e31c80-d476-11eb-8221-0359d92886d6.png)

**Reply by prologic** [reply]
That's a bit of a different response to what I we from them. 🤣

**Comment by jwillmer** [comment]
Whats the deal with this docker plugin to use SeaweedFS with docker? It seams it got no love since two years. @chrislusf is SeaweedFS intended to be used with docker or do you discourage from it? My current problem is that I need a distributed storage system that works with docker containers. I looked at GlusterFS, it does not have docker support but I can just bind mount the created folders. Not perfect but doable. How do you propose using SeaweedFS with docker if at all?

**Comment by jwillmer** [comment]
Did you notice [this repo](https://github.com/ggpwnkthx/docker-plugin-volume-seaweedfs)? Seams like it's the one with the most active development towards a docker plugin implementation.

**Reply by chrislusf** [reply]
`weed mount` or `weed fuse` should be simple to use with docker.  Need some expert to hook it up with docker.

The repo https://github.com/ggpwnkthx/docker-plugin-volume-seaweedfs from @ggpwnkthx used some recent features and should be up-to-date.

**Reply by ggpwnkthx** [reply]
Hey guys, I can't believe its only been 2 months since my last commit on that repo. I've switched jobs since then and haven't had a chance finish Docker Volume Plugin project for SeaweedFS. 

The trickiest part of getting SeaweedFS to work with Docker Swarm as well as a Docker Volume Plugin has less to do with SeaweedFS and more to do with how Docker handles networking. A Docker Volume Plugin **cannot** be attached to an Overlay Network (used by Docker Swarm containers). My work-around for this was to have the SeaweedFS containers use Unix sockets to tunnel the HTTP and gRPC traffic. Unfortunately, this means a proxy service has to be used to convert the IP traffic into Unix traffic and then back from Unix traffic into IP traffic.

I attempted to write my own [proxy service in go-lang](https://github.com/ggpwnkthx/go-proxet), and I've been half tempted to write a Docker Networking Plugin instead of using socat/nginx/haproxy. However, as much as that learning experience sounds awesome to me, I feel that might be overkill for this situation.

If SeaweedFS could be altered to allow for Unix socket(s) directly, the need for a proxy service (when using SeaweedFS in a Docker Swarm and as a Volume Plugin) can be removed. Additionally, the Volume Plugin that I wrote could be significantly simplified.

@chrislusf Any thoughts/opinions on the difficulty of adding Unix socket communication into SeaweedFS directly?

**Reply by chrislusf** [reply]
@ggpwnkthx thanks for your work and your insights to the docker networking!

Seems the work is to add unix socket support for filer? This should be fairly easy. We can listen on an optional unix socket around here: https://github.com/chrislusf/seaweedfs/blob/master/weed/command/filer.go#L217

Not sure whether the `weed mount`also need some changes.

Welcome for PR!

**Reply by prologic** [reply]
@ggpwnkthx Actually I _think_ your particular use-case is a bit different to mine. You are correct, Docker Plugins cannot access overlay networks, they operate at the host network. However if you have your SeaweedFS cluster running separate to your Swarm cluster (_regardless of how it's actually deployed, SeaweedFS itself could be deploy on a separate Swarm cluster_) then you can use the regular `weed mount` CLI by `os/exec` 'ing in the plugin itself and mounting the the SeaweedFS volume in the container's filesystem.

This is precisely what the [Docker Volume Plugin for SeaweedFS](https://github.com/seaweedfs/docker-plugin_seaweedfs) does, forked from https://github.com/katharostech/docker-plugin_seaweedfs and furhter updated in my work at https://github.com/prologic/docker-plugin_seaweedfs

So the complexity of what you've  built IHMO is a bit unnecessary with UNIX sockets, proxying and what not.

**Comment by nivekuil** [comment]
Here's the swarm service spec that I use to get `weed mount` working in a docker swarm. I run it globally, mounting from `/mnt/weed` on the host to `/docker/volumes` within seaweedfs. I generate the json directly, so no "pretty" yaml, but you can probably see what's going on. Essentially, you have to go docker-in-docker for `/dev/fuse` and then connect the inner container to a network that you can find both filer and volume on. @prologic  @jwillmer 
```json
{
  "mount" : {
    "image" : "docker:20-dind",
    "entrypoint" : "sh -c \"umount -l /mnt/weed || true; docker run --rm --security-opt apparmor:unconfined --cap-add SYS_ADMIN --net=global --device /dev/fuse -v /mnt:/mnt:rshared chrislusf/seaweedfs mount -dir=/mnt/weed -dirAutoCreate -filer=filer:8888 -filer.path=/docker/volumes\"",
    "cap_add" : [ "SYS_ADMIN" ],
    "volumes" : [ "/mnt:/mnt:rshared", "/var/run/docker.sock:/var/run/docker.sock:ro" ],
    "deploy" : {
      "mode" : "global"
    }
  }
}
```
Note that I only use this for non-critical things. I don't think `weed mount` is meant to be POSIX compliant.

**Comment by chrislusf** [comment]
@nivekuil Do you want to put this into a wiki similar to this https://github.com/chrislusf/seaweedfs/wiki/Deployment-to-Kubernetes-and-Minikube ? 
Should be useful to some use cases.

**Comment by atyurin80** [comment]
Seems to be useful to add in docker-compose.yml for 'master' service the option 'mode: global' here https://github.com/seaweedfs/seaweedfs/wiki/SeaweedFS-in-Docker-Swarm

```
services:
  master:
..
    deploy:
      mode: global
```

**Reply by Synertry** [reply]
Wow thank you for this, you made my week! I was already debugging the stack too deep, because the ip connection flags refer to the actual container name. (forgot that)
But then I noticed at max 1 volume server connecting to master.

Edit: Don't. See answer below.

**Reply by chrislusf** [reply]
Need a PR for this?

**Reply by Synertry** [reply]
No wait. I retract that my answer. I am still debugging. With the global mode I have 4 masters, with seperate IPs and have not specified the _peers_ flag yet. Does not seem right.

The single replica of master should suffice, shouldn't it? The volume instances except for one however cannot connect to the grpc port of the master. Either I have to configure my network overlay with further options or it's another problem which I have not found yet.

**Reply by Synertry** [reply]
After days of debugging and reading various documentation, it was the virtual network card _vmxnet3_ on our VMWare Server, which communicates over tcp 4789, the same default port for the data port of a swarm. Somehow the routing mesh gets confused with it.

As for OPs post. I would not activate set the global mode for the master. It works with 1 and 1 is enough as stated here: https://github.com/seaweedfs/seaweedfs/wiki/Production-Setup#one-master-is-fine
But thanks for checking by.
